### ISSUE TITLE
Should setting hydrogen masses in ForceField also subtract mass from parent heavy atom?

### ISSUE THREAD

**Issue by jchodera** [openmm/openmm#3435]
We've observed cases in which using too heavy of a hydrogen mass can cause too much mass to be subtracted from the parent heavy atom, causing instabilities instead of preventing them.

Instead, increasing the hydrogen mass _without_ subtracting the mass from the parent atom is likely to not introduce new instabilities, but rather stabilize the system, while at worst perhaps slowing diffusion a bit.

Is there a reason to prefer the current method over the alternative? Could we offer the alternative method, perhaps as default?
If data is needed to justify this, what experiment would demonstrate the greater utility of one method over the other?

**Comment by giadefa** [comment]
There was some indication that the total mass had to stay the same, but I
never quite understood why.
Clearly scaling all the masses up for every atom does not achieve anything.

**Comment by peastman** [comment]
If you increase the total mass, you slow down all motions, which is equivalent to rescaling time.  There's no benefit to that.  It doesn't help you sample any faster.  The goal of mass repartitioning is to slow down the fastest motions that limit the time step, without slowing down the large scale motions you're trying to study.

**Comment by jchodera** [comment]
Exactly. We would like to offer users ways to do this without introducing
new timestep limiting instabilities. Subtracting mass from the parent heavy
atom does not accomplish this goal.

Increasing hydrogen masses without subtracting mass might be equivalent to
rescaling time and subtracting mass from *all* heavy atoms. What if we did
that instead?

Instead of subtracting mass from the attached heavy atom, we subtract mass
from all heavy atoms to avoid introducing local issues?

**Comment by jchodera** [comment]
That's why I asked what experiments you would like to see. The issue here
is that we have repeatedly seen choices that lead to *instabilities* and
crashes. We don't want our users to run into this if there is a safer
alternative. So what would that safer alternative be, and what experiments
would demonstrate it's superiority?

**Comment by peastman** [comment]
If you're really interested in this as a project, start by reviewing the literature on the field, which is extensive and goes back decades, a least to [Bennett's classic 1975 paper](https://doi.org/10.1016/0021-9991(75)90077-7) on mass tensor MD.  A lot of ideas have been suggested for ways to modify masses to optimize sampling, and a lot of simulations have been run trying them out.  The repartitioning scheme used in OpenMM has been adopted as the standard method in the community because 1) it is simple to implement, 2) it doesn't add significant cost, 3) it has minimal effect on the kinetics of larger scale motions, and 4) it's effective at increasing the step size, as long as you don't try to do too much of it.

**Comment by jchodera** [comment]
@peastman: I am indeed familiar with the extensive literature on the topic,
and also the limitations of the experiments that were used to supposedly
demonstrate the superiority of one method over another (or at least the
appearance of minimal perturbation).

What I'm asking about is what we, with the benefit of half a century of
accumulated wisdom since the earliest papers here, would feel are
compelling experiments that would demonstrate a method without as many
pitfalls is generally more robust.

In order to demonstrate this convincingly, a new method should:
* Permit a timestep to be used over a larger Hmass range without
instabilities
* Provide an *efficiency* (defined as number of uncorrelated snapshots per
unit computer time) that is at least as good as the current approach for a
variety of test systems (say, from openmmtools.testsystems). The
statistical inefficiency of, say, the potential energy would be used to
quantify the number of uncorrelated samples, following [this work](
https://pubs.acs.org/doi/10.1021/acs.jctc.5b00784)

Is this a reasonable

**Comment by peastman** [comment]
I've actually had an idea about this at the back of my mind for years, but I've never gotten around to trying to implement it.  The time step is limited by the highest frequency motion in the system.  To optimize sampling, you ideally want every motion to have the same frequency.  In that case, no motion is limiting the sampling of any other motion.  Mass tensor MD tries to achieve that, but it has two disadvantages.  First, it's expensive, involving extra work on every step.  Second, because it globally redistributes mass, it alters the kinetics of large scale motions.  The latter isn't a problem if you only care about sampling, but it is if you care about kinetics.

Mass repartitioning fixes those problems by adding no overhead and by only doing local redistribution.  But it has two disadvantages.  First, it requires the user to pick a hydrogen mass, which has to be done by trial and error.  Second, because it sets every hydrogen to the same mass, it has less ability to optimize all motions at once.

My idea is to do local repartitioning in a more intelligent way.  Every atom could transfer a different amount of mass to its hydrogens.  That would be a free variable.  We would look at the bond and angle terms involving each heavy atom, determine the frequency of each one as a function of the mass transferred from each atom it involves, and optimize them to minimize the frequency of the fastest motion for each atom.

Regarding the criteria, that depends on the application.  There are two main cases.

1. Optimize sampling per time step.  This is relevant for applications where you only care about sampling.  The step size is not directly relevant here: if you increase the total mass, of course you can increase the step size, but that doesn't mean you're sampling more.
2. Maximize the step size while preserving the kinetics of large scale motions.  This is relevant for applications where you care about kinetics.
